phoenix-core/src/main/java/org/apache/phoenix/hbase/index/IndexRegionObserver.java

```diff
@@ -1165,4 +1356,255 @@ public static void enableIndexing(HTableDescriptor desc, Class<? extends IndexBu
-}
\ No newline at end of file
+
+    /**
+     * This function has been adapted from PhoenixIndexBuilder#executeAtomicOp().
+     * The critical difference being that the code in PhoenixIndexBuilder#executeAtomicOp()
+     * generates the mutations by reading the latest data table row from HBase but in order
+     * to correctly support concurrent index mutations we need to always read the latest
+     * data table row from memory.
+     * It takes in an atomic Put mutation and generates a list of Put and Delete mutations.
+     * The list will be empty in the case of ON DUPLICATE KEY IGNORE and the row already exists.
+     * In the case of ON DUPLICATE KEY UPDATE, we will generate one Put mutation and optionally
+     * one Delete mutation (with DeleteColumn type cells for all columns set to null).
+     */
+  private List<Mutation> generateOnDupMutations(BatchMutateContext context, Put atomicPut) throws IOException {
+      List<Mutation> mutations = Lists.newArrayListWithExpectedSize(2);
+      byte[] opBytes = atomicPut.getAttribute(ATOMIC_OP_ATTRIB);
+      if (opBytes == null) { // Unexpected
+          return null;
+      }
+      Put put = null;
+      Delete delete = null;
+      // We cannot neither use the time stamp in the Increment to set the Get time range
+      // nor set the Put/Delete time stamp and have this be atomic as HBase does not
+      // handle that. Though we disallow using ON DUPLICATE KEY clause when the
+      // CURRENT_SCN is set, we still may have a time stamp set as of when the table
+      // was resolved on the client side. We need to ignore this as well due to limitations
+      // in HBase, but this isn't too bad as the time will be very close the the current
+      // time anyway.
+      long ts = HConstants.LATEST_TIMESTAMP;
+
+      byte[] rowKey = atomicPut.getRow();
+      ImmutableBytesPtr rowKeyPtr = new ImmutableBytesPtr(rowKey);
+      // Get the latest data row state
+      Pair<Put, Put> dataRowState = context.dataRowStates.get(rowKeyPtr);
+      Put currentDataRowState = dataRowState != null ? dataRowState.getFirst() : null;
+
+      if (PhoenixIndexBuilder.isDupKeyIgnore(opBytes)) {
+          if (currentDataRowState == null) {
+              // new row
+              mutations.add(atomicPut);
+          }
+          return mutations;
+      }
+
+      ByteArrayInputStream stream = new ByteArrayInputStream(opBytes);
+      DataInputStream input = new DataInputStream(stream);
+      boolean skipFirstOp = input.readBoolean();
+      short repeat = input.readShort();
+      final int[] estimatedSizeHolder = {0};
+      List<Pair<PTable, List<Expression>>> operations = Lists.newArrayListWithExpectedSize(3);
+
+      // store the columns that need to be read in the conditional expressions
+      final Set<ColumnReference> colsReadInExpr = new HashSet<>();
+      while (true) {
+          ExpressionVisitor<Void> visitor = new StatelessTraverseAllExpressionVisitor<Void>() {
+              @Override
```

Comment:
nit: good place for an Extract Method, since the logic's complex and not needed to follow the logic of the larger method. Also good to stick an extra comment explaining the column parsing.

Reply:
@gjacoby126 good idea.